### ISSUE TITLE
Can't open configure dialog window in another thread?

### ISSUE THREAD

**Issue by jacktang** [ib-controller/ib-controller#135]
Hello,

I try to write ConfigureXXXTask like ConfigureApiSettingTask 
```
    @Override
    public void run() {
        try {
            Utils.logToConsole("Thread is running");
            // ConfigDialogManager.configDialogManager().clearConfigDialog();
            // blocks the thread until the config dialog is available
            final JDialog configDialog = ConfigDialogManager.configDialogManager().getConfigDialog();
            Utils.logToConsole("configDialog: " + configDialog);

            GuiExecutor.instance().execute(new Runnable(){
                @Override
                public void run() {
                    configure(configDialog, level);
                }
            });

        } catch (Exception e) {
            Utils.logError("" + e.getMessage());
        }
    }
```

and scheduled by 
```
MyCachedThreadPool.getInstance().execute(new ConfigureXXXTask(...));
```

But the configure dialog is not going to show up. And the logging is 
```
18:26:26:976 IBController: thread is running
```
I invoke clearConfigDialog() method before getting the config dialog and failed. Could you please tell me how to re-open config dialog in another thread? Thanks

**Comment by rlktradewright** [comment]
I've been investigating this, and there appears to be a race condition when two such tasks are both trying to access the Configuration Dialog. In fact there seem to be two problems:

1. Depending on timing, neither may succeed in getting the config dialog

2. If one task succeeds in getting the config dialog, the other will also get the same instance. However the first configuration routine to run makes its settings and then clicks ok and closes the dialog before the other routine starts, so the second one fails because the dialog structure has been dismantled.

This will need a bit of clear thinking to resolve, and I probably won't get it sorted today.

So just to make sure that your code is ok, can you try running yours again but first make sure that there is nothing else that's trying to open the config dialog. In particular make sure that `ForceTwsApiPort` is not set in IBController.ini, ie you want:

`ForceTwsApiPort=`

If you still have a problem when yours is the only thing that's trying to get the configuration dialog, then there must be some other reason.

**Comment by rlktradewright** [comment]
By the way, before you do too much work on this, can I remind you that there have been several pull requests in the past that wish to configure additional API settings, and they have all been rejected because IBController is not intended to be a TWS/Gateway configuration tool.

I realise that there are situations where people find this to be an awkward limitation (though people seem to be living with it ok), and the question is the best way to provide a solution without getting into an endless maintenance loop where more and more configuration options are requested. I think it would be a mistake to try to put together a pull request that tries to solve this problem until we've agreed on the best strategy. 

So if you're trying to code up a more generic API configuration mechanism that you intend to submit as a PR, be prepared for it to be rejected. I'm not saying that it will definitely be rejected - just that it'll need to be pretty darn convincing to be accepted!

**Comment by rlktradewright** [comment]
@jacktang 

I've fixed the issues identified earlier. So if you pull the latest to your local repository and try running your stuff again, it might help.

What happens now if that the creation of the future is properly synchronized so that only one future is created. And a usage count is kept of the number of 'things' (potentially including the user) accessing the config dialog, and the 'ok' button isn't clicked until they've all finished.

Theoretically if the user were quick enough to open the config dialog manually before such configuration tasks ran, he could keep the config dialog open indefinitely and prevent them from completing. However I can't see any practical way of the user achieving this.

**Comment by rlktradewright** [comment]
Addressed by Release 3.4.0
